src/main/java/org/openlmis/upload/TradeItemService.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * This program is part of the OpenLMIS logistics management information system platform software.
+ * Copyright © 2017 VillageReach
+ *
+ * This program is free software: you can redistribute it and/or modify it under the terms
+ * of the GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *  
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
+ * See the GNU Affero General Public License for more details. You should have received a copy of
+ * the GNU Affero General Public License along with this program. If not, see
+ * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
+ */
+
+package org.openlmis.upload;
+
+import static org.openlmis.upload.RequestHelper.createUri;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonObjectBuilder;
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestClientResponseException;
+
+@Service
+public class TradeItemService extends BaseCommunicationService {
+
+  private static final String PRODUCT_CODE = "productCode";
+  private static final String IDENTIFIERS = "identifiers";
+  private static final String TRADE_ITEM = "tradeItem";
+
+  @Autowired
+  private OrderableService orderableService;
+
+  @Getter(AccessLevel.PUBLIC)
+  private Map<String, String> cachedTradeItemIdByOrderableCode;
+
+  @Override
+  public void before() {
+    JsonArray orderables = this.orderableService.findAll();
+    cachedTradeItemIdByOrderableCode = IntStream.range(0, orderables.size())
```

Comment:
JsonArray extends Collection interface so it has the stream method.

Reply:
The stream method returns a stream with JsonValue elements. With this in mind, I think would have to cast somehow to JsonObject to perform filter and get the key values for orderable code and trade item id.